File: bindings/swift/Sources/DataRCT/Discovery.swift
//
//  Discovery.swift
//  
//
//  Created by Julian Baumann on 06.01.24.
//

import Foundation

public  protocol DiscoveryDelegate: DeviceListUpdateDelegate {
    func discoveryDidUpdateState(state: BluetoothState)
}

public class Discovery {
    private let internalHandler: InternalDiscovery
    private let bleImplementation: BLEClientManager
    
    public init(delegate: DiscoveryDelegate) throws {
        internalHandler = try InternalDiscovery(delegate: delegate)
        bleImplementation = BLEClientManager(delegate: delegate, internalHandler: internalHandler)
        internalHandler.addBleImplementation(implementation: bleImplementation)
    }
    
    public func startScan() throws {
        try bleImplementation.ensureValidState()
        
        internalHandler.start()
    }
    
    public func stopScan() throws {
        try bleImplementation.ensureValidState()
        
        internalHandler.stop()
    }
}

File: bindings/csharp/DataRct/Discovery.cs
using DataRct.BluetoothLowEnergy;

namespace DataRct;

public class Discovery
{
    private readonly InternalDiscovery _internalHandler;
    private readonly BleClient _bleImplementation;

    public Discovery(DiscoveryDelegate discoveryDelegate)
    {
        _internalHandler = new InternalDiscovery(discoveryDelegate);
        _bleImplementation = new BleClient(discoveryDelegate, _internalHandler);
        
        _internalHandler.AddBleImplementation(_bleImplementation);
        
    }

    public void StartScan()
    {
        _internalHandler.Start();
    }

    public void StopScan()
    {
        _internalHandler.Stop();
    }
}